IIIF manifest block

dannylamb opened the pull request:
**GitHub Issue**: https://github.com/Islandora-CLAW/CLAW/issues/932

# What does this Pull Request do?

Adds a block you can place using Context that uses IIIF manifests to render a multi-page openseadragon viewer.

# What's new?
A new type of block, which takes a IIIF manifest url (or a url pattern using tokens) as configuration.

Tests.

# How should this be tested?

- Set your openseadragon settings to use 'Sequence Mode' if you want multiple pages.
- Make a view that renders manifests using our IIIF Manifest views style plugin (or just use [this](https://gist.github.com/dannylamb/c90686957d3d5599f49f2cab02f045d4))
- Make a block using the block placement screen and enter a manifest URL in the config
- If you really wanna go for it, place the block using context when the node has the 'Paged Content' tag, and use `http://localhost:8000/node/[node:nid]/book-manifest` as the IIIF Manifest url.
  - Remember, you can adjust weights between the core and context block placement UIs to put that block wherever you want, like slighly above the main content but below the tabs, for example.
- Behold your multi-page openseadragon viewer rendered using a IIIF manifest!

# Additional Notes:
We'll need additional default config in order to pin down the manifest view and the context, but you get the picture.

# Interested parties
@Islandora-CLAW/committers @dbernstein @elizoller @rangel35

elizoller commented:
I pulled in this branch, have a paged content object, the book-manifest URL, set OSD to sequence mode, added http://localhost:8000/node/[node:id]/book-manifest to the block settings and placed the block via context.
I can see the block (the title is present) but no viewer.
Did I miss a step?
I'm getting a "cannot read property 'options' of undefined' on line 28 of the js file: ` base = '#' + settings.openseadragon.options.id;`

dannylamb commented:
@elizoller Anything in the watchdog logs?

whikloj commented on src/Plugin/Block/OpenseadragonBlock.php:
```diff
@@ -0,0 +1,63 @@
+    $form['iiif_manifest_url'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('IIIF Manifest URL'),
+      '#description' => $this->t('URL of the IIIF manifest to render.  You may use tokens to provide a pattern (e.g. "http://localhost/node/[node:id]/manifest"'),
```
End of this line
> (e.g. "http://localhost/node/[node:id]/manifest"')

should be
> (e.g. "http://localhost/node/[node:nid]/manifest")')

elizoller replied:
This fix worked for me.

dannylamb commented:
@whikloj @elizoller I've updated the config page to have a somewhat more meaningful description.

whikloj commented on src/Plugin/Block/OpenseadragonBlock.php:
```diff
@@ -0,0 +1,63 @@
+    $form['iiif_manifest_url'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('IIIF Manifest URL'),
+      '#description' => $this->t('Absolute URL of the IIIF manifest to render.  You may use tokens to provide a pattern (e.g. "http://localhost/node/[node:nid]/manifest"'),
```
You are still missing the closing (in the text) bracket here.
It displays like
> ... You may use tokens to provide a pattern (e.g. "http://localhost/node/[node:nid]/manifest"

whikloj replied:
I have some code to list tokens here like it does for the title field if you so desire it. Or I can add that as a separate PR.

dannylamb replied:
if it's the same as scraping code from core to display the token list... i can hunt that down if you don't want it coming from you.

whikloj replied:
I'll just leave this here 😉 
https://gist.github.com/whikloj/4fa01684dcbf50a1b75f38298ca80a68

dannylamb commented:
Ha, missed your last reply but ended up doing something very similar.  Let me know what you think @whikloj

dannylamb commented:
@whikloj This is good to go now, I think.

whikloj commented:
Soooo, this definitely works. I like providing the token help, but now I am worried that most of these tokens are not relevant.

As it displays now you get these options:
![Screen Shot 2019-09-20 at 8 40 11 AM](https://user-images.githubusercontent.com/2857697/65333611-74038880-db86-11e9-9db8-33bd9a9f6b86.png)

So you can remove those `current-date`, `current-user`, `random` by adding `"#global_types" => FALSE,` to your token_tree_link array.

That results in only displaying node tokens.
![Screen Shot 2019-09-20 at 9 05 52 AM](https://user-images.githubusercontent.com/2857697/65333775-b88f2400-db86-11e9-95dd-3c7495f60ea5.png)

This is reasonable, but _now_ I am wondering if providing all these tokens going to make people cause themselves more trouble?

I don't want to hold this up...it works so I'll throw this here for some quick discussion and perhaps the resolution is to have people look at the UI/UX later.

dannylamb commented:
I can do that real quick @whikloj

dannylamb commented:
done

dannylamb commented:
Those tokens are there for one in particular, `[node:nid]`.  I don't know why you'd want to use others, we just need that one to line up with the same pattern you put into the manifest view.  I think documentation would go a long way.

whikloj commented:
Waiting on Travis
![Cookie waiting](https://media.giphy.com/media/o5oLImoQgGsKY/giphy.gif)

dannylamb commented:
Ah... @Natkeeran may have just found a bug i introduced wtih the token help text.

dannylamb commented:
Thanks to @Natkeeran for finding that out and notifying us on Slack.  I nested a the field in a fieldset and forgot i'd have to hack out the manifest url a little differently when saving.

:disappointed_relieved: Should work now.

whikloj commented on src/Plugin/Block/OpenseadragonBlock.php:
```diff
@@ -0,0 +1,70 @@
+   * {@inheritdoc}
+   */
+  public function blockSubmit($form, FormStateInterface $form_state) {
+    $this->configuration['iiif_manifest_url'] = $form_state->getValues()['iiif_manifest_url_fieldset']['iiif_manifest_url'];
```
Seemingly the correct form of this is
```
$form_state->getValue(['iiif_manifest_url_fieldset', 'iiif_manifest_url']);
```
As it uses [NestedArray::getValue](https://api.drupal.org/api/drupal/core!lib!Drupal!Component!Utility!NestedArray.php/function/NestedArray%3A%3AgetValue/8.7.x)

Finding the syntax was a little bit of duckduckgo'ing

dannylamb replied:
wow... i had no idea you could do that...

I'll update it.  Also, found that there's a validation function i can toss in, too.

dannylamb commented:
@whikloj Ok, updated that.  Added the validation function too.  So now if you enter `http://localhost:8000/node/[node:notatokenwhatsoever]/book-manifest` it'll barf on you.
